totalRP3/UI/Tooltip/Tooltip.lua

```diff
@@ -0,0 +1,473 @@
+end
+
+local function UnpackColor(color)
+	if color then
+		return color.r, color.g, color.b;
+	else
+		return NORMAL_FONT_COLOR:GetRGB();
+	end
+end
+
+local function UnpackTooltipPadding(left, right, top, bottom)
+	return right, bottom, left, top;
+end
+
+local function ProcessLineDescription(tooltip, description)
+	if not description:ShouldShow() then
+		return;
+	end
+
+	local leftText = description:GetLeftText();
+	local rightText = description:GetRightText();
+
+	if leftText == "" and rightText == "" then
+		tooltip:AddLine(" ");
+	elseif rightText == "" then
+		local color = description:GetLeftTextColor();
+		local r, g, b = UnpackColor(color);
+		local wrap = description:ShouldWordWrap();
+		local leftOffset = description:GetLeftOffset();
+
+		if color and description:IsColorWrappingEnabled() then
+			leftText = color:WrapTextInColorCode(leftText);
+		end
+
+		tooltip:AddLine(leftText, r, g, b, wrap, leftOffset);
+	else
+		local leftColor = description:GetLeftTextColor();
+		local leftR, leftG, leftB = UnpackColor(leftColor);
+		local rightColor = description:GetRightTextColor();
+		local rightR, rightG, rightB = UnpackColor(rightColor);
+		local wrap = description:ShouldWordWrap();
+		local leftOffset = description:GetLeftOffset();
+
+		if leftColor and description:IsColorWrappingEnabled() then
+			leftText = leftColor:WrapTextInColorCode(leftText);
+		end
+
+		if rightColor and description:IsColorWrappingEnabled() then
+			rightText = rightColor:WrapTextInColorCode(rightText);
+		end
+
+		tooltip:AddDoubleLine(leftText, rightText, leftR, leftG, leftB, rightR, rightG, rightB, wrap, leftOffset);
+	end
+
+	local leftFontString, rightFontString = TRP3_TooltipUtil.GetLineFontStrings(tooltip, tooltip:NumLines());
+	local leftFontObject = description:GetLeftFontObject();
+	local rightFontObject = description:GetRightFontObject();
+
+	-- Font objects can't be assigned until lines are added, but doing so
+	-- also means that any previously used colors are dropped if the assigned
+	-- font object would change - hence the need for SetTextColor calls here.
+
+	leftFontString:SetFontObject(leftFontObject);
+	leftFontString:SetFixedColor(description:ShouldUseFixedColor());
+	leftFontString:SetTextColor(UnpackColor(description:GetLeftTextColor()));
+	rightFontString:SetFontObject(rightFontObject);
+	rightFontString:SetFixedColor(description:ShouldUseFixedColor());
+	rightFontString:SetTextColor(UnpackColor(description:GetRightTextColor()));
+end
+
+function TRP3_Tooltip.ProcessTooltipDescription(tooltip, description)
+	if not description:ShouldShow() then
+		return;
+	end
+
+	local owner = description:GetOwner();
+	local anchor = description:GetAnchor();
+
+	if type(anchor) ~= "table" then
+		tooltip:SetOwner(owner, anchor, description:GetAnchorOffset());
+	else
+		tooltip:ClearAllPoints();
+		tooltip:SetPoint(anchor:Get());
+		tooltip:SetOwner(owner, "ANCHOR_PRESERVE");
+	end
+
+	tooltip:SetOwner(owner, "ANCHOR_RIGHT");
```

Comment:
why are we setting the owner again after setting it right before?

Reply:
you saw nothing